### PR TITLE
Add CT-CODE support for the "In Order" and "Random" VS settings

### PR DESCRIPTION
**Pull request by DimitriPilot3**
Some of the routines that update the ```currentCourse``` variable are left unpatched in CT-CODE.  Two of them are related to the "In Order" and "Random" VS course selection settings, and are encountered when "Next Course" is selected (at the end of a race), or when "Random" VS chooses its first course at random (skipping the course screen altogether).  Not patching those would result in repeating the last played course over and over, since they touch the original ```currentCourse``` variable instead of CTGP's.

Additionally, both "In Order" and "Random" VS generate a pool of up to 32 course IDs prior to starting, which could be enhanced by allowing custom tracks to be pooled.  Since the "Random" pool is basically an "In Order" pool followed with random permutations, and since it'd make sense for "Random" VS to only pick custom tracks, the last CT cup should wrap back to the first CT cup instead of the Mushroom cup.  For completion's sake, the Lightning Cup (8.4: rBC64) should wrap back to the Mushroom Cup (1.1: Luigi Circuit) in case the pool has started off with a Nintendo course.  This effectively means *making the whole pool exclusive to one category*, CTs or Nintendo courses, depending on what the first picked course is.

This pull request addresses all of the above notes.  I started off by making patches for **PAL** (debugged on Dolphin and tested on my Wii U), and proceeded to *auto-port* the offsets to other regions using the "Mario Kart Wii Code Generator" program, and common sense to ensure the mods are region-agnosic.  **When testing other regions, make sure that "In Order" wraps correctly, and that nothing crashes.**

*Note that two CTs that are more than 32 slots away from each other (with wrapping taken into consideration) will never be played together in a Random VS pool due to its limited capacity (up to 32), unless the algorithm were changed to something better.*

*For memory-debugging purposes, look around at [menuData + 0x98] + 0x78 for the course pool's array.*

### CODE REVIEW

**Comment by Leseratte10** [comment]
Hello DimitriPilot3, 

some other questions to you (as you seem to be one of the few who is able to read / understand / write own assembler code besides Chadderz & MrBean) about bugs in 1.02: 

(I haven't tested your modifications yet so sorry if this already solves some of the things I write here): 

- This also makes online "random" work (also select custom track)?
-- If yes: When is the pool reset? At the beginning of a GP? Or for each race?
-- If not: Can you make it work, or at least make it not vote random for a whole GP (with the current random function)?
- Are you able to fix the online course selection so it actually focuses the last voted Cup and track and not the last played Cup?
- Are you able to modify the blue shell drag patch so the dragging only works offline, in regions, and in friend rooms but not in global races?

It would be great if you'd answer my questions. 

Leseratte10

**Comment by Chadderz121** [comment]
Hello @DimitriPilot3,

First of all, thanks a lot for the patch, I'm amazed anyone took the time to understand the source and actually fix something :wink:

I'll give it a test on my Wii, and test the port when I get chance and then merge this request.

@Leseratte10:
 1. @mrbean35000vr fixed that random friend room bug in v1.03 last week, so I'll try to backport that fix at the same time as I merge this.
 2. Never looked into that, I'd be more than happy to accept a fix for that if anyone can be bothered to write one.
 3. Currently the blue shell's item behaviour in an item behaviour table is modified directly by this mod. Instead, the value would have to be swapped in with the correct value as modes were changed. Not impossible, but a non-trivial change, as we don't already have code which executes at the right time.

Chadderz

**Comment by Leseratte10** [comment]
Aren't the assembler codes basically the same for both 1.02 and 1.03 so you could just copy the module? Or is CTGP messing with the RAM and actually moving memory parts around so the offsets aren't identical to vanilla mkw?

**Comment by Chadderz121** [comment]
They are basically the same, but v1.03 is still (for some reason) compiled by hand; it doesn't use this build system. I effectively disassemble the v1.03 code to work out what to put here. This build system is also parametrised, whereas v1.03 makes fixed choices, so I have to check the parameters don't interfere. It's not too difficult, but it's not just copy/paste.

File: bad1/bad1Data/mod/track_redirect.mod [hunk 1]
```diff
@@ -422,6 +422,142 @@ MOD_REL(
...
+ * For reference: "Random" picks the 1st course, pools 31 more in subsequent order,
+ * then randomly permutes each pooled course with another. Consequently, on CT-CODE,
+ * courses that are 32+ slots apart from each other are never pooled together.
+ */
+MOD_REL(
+    mod_redirect_vs_1,
+    mod_redirect_vs_1_addr,
+        /* Reload current track ID into param r4. "In Order" passes us its slot instead. */
+        lis r6, raceCupTable@ha;
+        lwz r4, currentCourse@l(r6);
```
Comment (DimitriPilot3) [inline bad1/bad1Data/mod/track_redirect.mod]: Now that I think about it, the mods that redirect ```raceCupTable``` (vs_1 twice, vs_6, vs_7) could be isolated, renamed, and moved up into the section of the file that groups all the ```mod_raceCupTable``` mods together.
Once that's done, all that would be left from this mod (besides redirection) is this particular instruction. Heck, it wouldn't be needed at all, if not for one of the callees of this function passing into r4 a special slot ID that's "returned" by ```redirect_course```.

File: bad1/bad1Data/mod/track_redirect.mod [hunk 2]
```diff
@@ -422,6 +422,142 @@ MOD_REL(
...
+        stw r0, 52(r1);
```
Comment (DimitriPilot3) [inline bad1/bad1Data/mod/track_redirect.mod]: Oh, and this instruction is a copy of existing code.

Comment (DimitriPilot3) [inline bad1/bad1Data/mod/track_redirect.mod]: Could also be omitted since it is a recopy of existing game code.

**Comment by Wiimm** [comment]
@DimitriPilot3
@Chadderz121 
Today I tested this patch for all 3 regions. I played offline 3 races in-order and 3 random at each region. And it worked like DimitriPilot3 told. In the next days I'll test, if it have online side effects.

**Comment by Wiimm** [comment]
Can't find any side effects if playing online. We tested it with PAL and USA versions. I don't see any reason against the merge.

**Comment by Chadderz121** [comment]
Thanks for the patch, I've now merged it into the main tree. Sorry it took so long!

**Pull request merged**